mbuild/utils/orderedset.py

```diff
@@ -0,0 +1,139 @@
+"""Ordered set module.
+
+Based on oset by Raymond Hettinger, with the following license:
+
+Copyright (c) 2009, Raymond Hettinger, and others All rights reserved.
+
+Package structured based on the one developed to odict Copyright (c) 2010,
+    BlueDynamics Alliance, Austria
+
+ - Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+ - Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+ - Neither the name of the BlueDynamics Alliance nor the names of its
+   contributors may be used to endorse or promote products derived from this
+   software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY BlueDynamics Alliance AS IS AND ANY EXPRESS OR
+IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+EVENT SHALL BlueDynamics Alliance BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+from collections.abc import MutableSet
+
+KEY, PREV, NEXT = range(3)
+
+
+class OrderedSet(MutableSet):
+    """An ordered set object with additional convenience methods.
+
+    Based on oset by Raymond Hettinger (see module for license)
+
+    Methods
+    -------
+    add
+    discard
+    index
+    list
+    pop
+    """
+
+    def __init__(self, iterable=None):
+        self.end = end = []
+        end += [None, end, end]  # sentinel node for doubly linked list
+        self.map = {}  # key --> [key, prev, next]
+        self._list = None
+        if iterable is not None:
+            self |= iterable
+
+    def __len__(self):
+        """Return the length."""
+        return len(self.map)
+
+    def __contains__(self, key):
+        """Determine whether the set contains a key."""
+        return key in self.map
+
+    # @property
```

Comment:
Do we need these commented-out lines? If not, can we remove them?

Reply:
no, haha, sorry I should mark this PR as WIP